### PR TITLE
Don't emit CA1849 for DbContext methods

### LINKED ISSUE #6684: CA1849 "Call async methods when in an async method" false positive on DbContext.AddRange()
### Analyzer

**Diagnostic ID**: [CA1849](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1849): `Call async methods when in an async method`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 7.0.400-preview.23274.1](https://dotnet.microsoft.com/download/dotnet/7.0)

### Describe the bug

Using `DbContext.AddRange(params object[])` reports a warning and suggests using `DbContext.AddRangeAsync(params object[])`.

However the `AddRangeAsync()` method exists only for very special edge cases and should not be used by default, as described on https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.dbcontext.addrangeasync?view=efcore-7.0

> This method is async only to allow special value generators, such as the one used by 'Microsoft.EntityFrameworkCore.Metadata.SqlServerValueGenerationStrategy.SequenceHiLo', to access the database asynchronously. For all other cases the non async method should be used.

### Steps To Reproduce

Use `DbContext.AddRange(params object[])` in an `async` method.

### Expected behavior

The rule is not triggered.

### Actual behavior

The rule is triggered.

## Additional context

The rule does not trigger in `IDbContextFactory<TContext>.CreateDbContext()` although `IDbContextFactory<TContext>.CreateDbContextAsync()` exists which is correct, so there already seems to be an exception for this very simmilar case.

### PR DESCRIPTION
**Pull request by CollinAlpert**
This PR adds exclusions for the `DbContext.Add()` and `DbContext.AddRange()` methods and fixes a false negative for `IDbContextFactory<DbContext>.CreateDbContext()` in the `UseAsyncMethodInAsyncContext` analyzer.

Fixes #6684.

### CODE REVIEW

**Comment by CollinAlpert** [comment]
The build is complaining that there is a missing entry in `RulesMissingDocumentation.md` for rule `CA2200`, however documentation does exist for that rule. Is this a bug in the build?

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseAsyncMethodInAsyncContext.cs [hunk 1]
```diff
@@ -162,29 +163,46 @@ public SyncBlockingSymbol(string Name, string Namespace, SymbolKind Kind, ISymbo
...
+                    .Concat(dbContextType
+                        .GetMembers("Add")
+                        .OfType<IMethodSymbol>())
+                    .Concat(dbContextType
+                        .GetMembers("AddRange")
+                        .OfType<IMethodSymbol>());
+            }
+
+            return methods;
+        }
```
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseAsyncMethodInAsyncContext.cs:204]: Lets use `ImmutableArray<IMethodSymbol>` instead
```suggestion
        private static ImmutableArray<IMethodSymbol> GetExcludedMethods(WellKnownTypeProvider wellKnownTypeProvider)
        {
            var methods = ImmutableArray.CreateBuilder<IMethodSymbol>();
            if (wellKnownTypeProvider.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftEntityFrameworkCoreDbContext, out INamedTypeSymbol? dbContextType))
            {
                foreach (var method in dbContextType.GetMembers().OfType<IMethodSymbol>())
                {
                    if (method.Name is "Add" or "AddRange")
                    {
                        methods.Add(method);
                    }
                }
            }

            return methods.ToImmutable();
        }
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseAsyncMethodInAsyncContext.cs [hunk 2]
```diff
@@ -56,41 +55,43 @@ public override void Initialize(AnalysisContext context)
...
                     return;
                 }
 
-                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemObsoleteAttribute, out INamedTypeSymbol? systemObsoleteAttribute))
+                if (!wellKnownTypeProvider.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemObsoleteAttribute, out INamedTypeSymbol? systemObsoleteAttribute))
                 {
                     return;
                 }
 
+                ISet<IMethodSymbol> excludedMethods = GetExcludedMethods(wellKnownTypeProvider).ToSet();
```
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseAsyncMethodInAsyncContext.cs:86]:
```suggestion
                ImmutableArray<IMethodSymbol> excludedMethods = GetExcludedMethods(wellKnownTypeProvider);
```

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseAsyncMethodInAsyncContextTests.cs [hunk 3]
```diff
@@ -1243,6 +1244,64 @@ End Module
...
+using Microsoft.EntityFrameworkCore;
+using System.Threading.Tasks;
+
+class Test {
+    public async Task RunAsync(IDbContextFactory<DbContext> factory) {
+        var context = {|#0:factory.CreateDbContext()|};
+    }
+}",
+                ExpectedDiagnostics = { new DiagnosticResult(UseAsyncMethodInAsyncContext.Descriptor).WithLocation(0).WithArguments("IDbContextFactory<DbContext>.CreateDbContext()", "IDbContextFactory<DbContext>.CreateDbContextAsync(CancellationToken)") },
+                ReferenceAssemblies = ReferenceAssemblies.Net.Net70.WithPackages(ImmutableArray.Create(new PackageIdentity("Microsoft.EntityFrameworkCore", "7.0.8")))
```
Comment (buyaa-n) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseAsyncMethodInAsyncContextTests.cs:1301]: Lets use static readonly field for `ImmutableArray.Create(new PackageIdentity("Microsoft.EntityFrameworkCore", "7.0.8")` and reuse for each test

**Comment by buyaa-n** [comment]
> The build is complaining that there is a missing entry in `RulesMissingDocumentation.md` for rule `CA2200`, however documentation does exist for that rule. Is this a bug in the build?

Strange, usually it complains when there is an entry that needs to be removed because the corresponding doc is added.

Anyway, better to run the `msbuild /t:pack` to see if there were any doc update happened

**Comment by CollinAlpert** [comment]
I ran that command already and no changes were made, which makes sense. I can try again tomorrow.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseAsyncMethodInAsyncContext.cs [hunk 4]
```diff
@@ -162,29 +163,46 @@ public SyncBlockingSymbol(string Name, string Namespace, SymbolKind Kind, ISymbo
...
+                var addMethods = dbContextType
+                    .GetMembers("Add")
+                    .OfType<IMethodSymbol>();
+                var addRangeMethods = dbContextType
+                    .GetMembers("AddRange")
+                    .OfType<IMethodSymbol>();
+                methods = methods.AddRange(addMethods.Concat(addRangeMethods));
+            }
+
+            return methods;
```
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseAsyncMethodInAsyncContext.cs:203]: NIT: ImmutableArray.Builder is more efficient to use
```suggestion
            var methodsBuilder = ImmutableArray.CreateBuilder<IMethodSymbol>();
            if (wellKnownTypeProvider.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftEntityFrameworkCoreDbContext, out INamedTypeSymbol? dbContextType))
            {
                foreach (var method in dbContextType.GetMembers().OfType<IMethodSymbol>())
                {
                    if (method.Name is "Add" or "AddRange")
                    {
                        methodsBuilder.Add(method);
                    }
                }
            }
            return methodsBuilder.ToImmutable();
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseAsyncMethodInAsyncContext.cs [hunk 5]
```diff
@@ -162,29 +163,45 @@ public SyncBlockingSymbol(string Name, string Namespace, SymbolKind Kind, ISymbo
...
+            if (wellKnownTypeProvider.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftEntityFrameworkCoreDbContext, out INamedTypeSymbol? dbContextType))
+            {
+                foreach (var method in dbContextType.GetMembers().OfType<IMethodSymbol>())
+                {
+                    if (method.Name is "Add" or "AddRange")
+                    {
+                        methodsBuilder.Add(method);
+                    }
+                }
+            }
```
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseAsyncMethodInAsyncContext.cs:201]:
```suggestion
            }
            
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseAsyncMethodInAsyncContext.cs [hunk 6]
```diff
@@ -162,29 +163,46 @@ public SyncBlockingSymbol(string Name, string Namespace, SymbolKind Kind, ISymbo
...
+
```
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseAsyncMethodInAsyncContext.cs:202]:
```suggestion
            }

```

**Review by buyaa-n — APPROVED** [review]
Thank you @CollinAlpert!

**Comment by molinch** [comment]
Do we know if this is already shipped with a .NET8 preview? Or it will land in next one?

**Comment by cremor** [comment]
Will this only ship with a .NET 8 SDK, or will it also be included in a .NET 7 SDK update?

**Comment by buyaa-n** [comment]
It should have included in .NET 8 preview 7, it is not ported to .NET 7

**Comment by Spacefish** [comment]
I get a lot of CA1849 warnings with .NET 8 Preview 7 for my DbSet.Add(..) calls.. 
As we are not calling the "Add" directly on the DbContext but instead on the Sets on the Contexts.. Maybe we should include that as well.

**Pull request merged**